## SQL

The coronavirus took the entire world by surprise, changing everyone's daily routine. City dwellers no longer spent their free time outside, going to cafes and malls; more people were at home, reading books. That attracted the attention of startups that rushed to develop new apps for book lovers. 

The goal of this study is to generate a value proposition for a new product base on the services that is competing in this market. 

In [4]:
import pandas as pd
import numpy as np
import math
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.dates
import seaborn as sns
import matplotlib.dates as mdates
import datetime 
import numpy as np
from sqlalchemy import create_engine

In [5]:
db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

<p><a name="Step 1"></a></p>

### Step 1. Download and check the data 

In [6]:
# make a query to download book dataset
books = pd.io.sql.read_sql('''SELECT * FROM books''', con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [7]:
#check the shape of the dataset
books.shape

(1000, 6)

In [8]:
#study the dataset by calling the info
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


The books dataset contains 1000 rows and 6 columns.Every column is in the correct format except for publication date which is in object format instead of datetime64 format.

In [9]:
# make a query to download the authors dataset
authors = pd.io.sql.read_sql('''SELECT * FROM authors''', con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [10]:
#find the shape of the dataset
authors.shape

(636, 2)

In [11]:
# call the info to study the dataset
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


authors dataset has 636 rows and 2 columns. columns are in correct datatype.

In [12]:
# make a query to download the publishers dataset
publishers = pd.io.sql.read_sql('''SELECT * FROM publishers''', con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [13]:
# check the shape of the dataset
publishers.shape

(340, 2)

In [14]:
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


publishers dataset has 340 rows and 2 colums, the publishers_id and publisher.All column are in correct format and no missing values.

In [12]:
# make a query to download the publishers dataset
ratings = pd.io.sql.read_sql('''SELECT * FROM ratings''', con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [13]:
#check the shape of the dataset
ratings.shape

(6456, 4)

In [14]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


ratings dataset has 6456 rows and 4 columns namely the rating_id, book_id, username, and rating,  there is no null values and all are in correct data type.

In [15]:
# make a query to download the publishers dataset
reviews = pd.io.sql.read_sql('''SELECT * FROM reviews''', con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [16]:
reviews.shape

(2793, 4)

In [17]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


reviews dataset has 2,793 rows and 4 columns namely: the review_id, book_id, username and text. there are no missing values and and the columns are in correct datatype.

<p><a name="Step 2"></a></p>

### Task

#### Find the number of books released after January 1, 2000.

In [18]:
query = ''' SELECT COUNT(*) FROM books WHERE publication_date > '2000-01-01' '''
result = pd.io.sql.read_sql(query, con = engine)
print('There are', result['count'][0] , 'number of books released after January 1, 2000.')

There are 819 number of books released after January 1, 2000.


#### Find the number of user reviews and the average rating for each book.

In [19]:
query2 = ''' SELECT 
                books.title AS Title,
                COUNT(reviews.review_id) as num_reviews,
                AVG(ratings.rating) as avg_rating   
            FROM books
            INNER JOIN ratings on ratings.book_id = books.book_id
            INNER JOIN reviews on reviews.book_id = books.book_id
            GROUP BY books.book_id
            ORDER BY num_reviews DESC'''

result2 = pd.io.sql.read_sql(query2, con = engine)
result2.head(10)

title  num_reviews  avg_rating
0                            Twilight (Twilight  #1)         1120    3.662500
1                The Hobbit  or There and Back Again          528    4.125000
2                             The Catcher in the Rye          516    3.825581
3  Harry Potter and the Prisoner of Azkaban (Harr...          492    4.414634
4  Harry Potter and the Chamber of Secrets (Harry...          480    4.287500
5               Angels & Demons (Robert Langdon  #1)          420    3.678571
6  Harry Potter and the Order of the Phoenix (Har...          375    4.186667
7  The Lightning Thief (Percy Jackson and the Oly...          372    4.080645
8  The Fellowship of the Ring (The Lord of the Ri...          370    4.391892
9                                        Animal Farm          370    3.729730

The book with the highest number of review is Twilight #1 though it was not the book with the highest number of rating.

#### Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).

In [20]:
query3 = ''' SELECT 
                publishers.publisher AS publisher,
                COUNT(DISTINCT books.book_id) as num_books   
            FROM books
            INNER JOIN publishers on publishers.publisher_id = books.publisher_id
            WHERE books.num_pages > 50
            GROUP BY publisher
            ORDER BY num_books DESC'''

result3 = pd.io.sql.read_sql(query3, con = engine)
result3.head(5)

publisher  num_books
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25
3          Penguin Classics         24
4                    Bantam         19

The number above shows the publisher that has released the greatest number of books with more than 50 pages.

#### Find the average number of text reviews among users who rated more than 50 books. 

In [77]:
query4 = ''' SELECT 
                reviews.username AS user_name,
                AVG(DISTINCT reviews.review_id) as Avg_review 
            FROM reviews
            INNER JOIN ratings on ratings.book_id = reviews.book_id
            GROUP BY user_name
            HAVING COUNT(ratings.book_id)>50
            ORDER BY user_name DESC'''

result4 = pd.io.sql.read_sql(query4, con = engine)
result4.head(5)

user_name   avg_review
0             zross  1232.200000
1           znelson  1133.750000
2         zjohnston  1268.857143
3            yweeks  1460.266667
4  yvonnevillarreal  1549.538462

The table above shows  the average number of text reviews among users who rated more than 50 books.